### PR TITLE
Added a best-practices section for checkpointing and future topics

### PR DESCRIPTION
**Pull request by marianalevineu**
Added a checkpointing section with examples.

### CODE REVIEW

**Comment by marianalevineu** [comment]
@ShobanaSekar  looks like you're already a reviewer. I'll fix some things today first and then will notify when it's done.

**Comment by marianalevineu** [comment]
@tketchem14 Added you as a reviewer.

File: docs/source/best-practices/checkpointing.rst [hunk 1]
```diff
@@ -0,0 +1,160 @@
...
+   
+  sbatch submit_tf_array.bash
+
+Checkpointing using DMTCP
+~~~~~~~~~~~~~~~~~~~~~~~~~~~~
+
+`DMTCP <https://dmtcp.sourceforge.io/>`_ (Distributed MultiThreaded Checkpointing) is a Checkpointing tool that lets you Checkpoint without the need to change your code. It Works with most Linux applications such as Python, Matlab, R, GUI, MPI etc. 
+The program runs in the background of your program, without significant performance loss, and saves the process states into checkpoint files.
+
+DMTCP is available on the cluster::
```
Comment (marianalevineu) [inline docs/source/best-practices/checkpointing.rst:133]: fix to show as code

File: docs/source/best-practices/checkpointing.rst [hunk 2]
```diff
@@ -0,0 +1,162 @@
+*****************************************
+Checkpoint/Restart Discovery Jobs
+*****************************************
+
+The complexity of HPC systems may introduce uncontrollable and unpredictable behaviors and may result in job failures due to hardware or software. Applying fault tolerance techniques in your HPC workflows allows your jobs to become more resilient to crashes, prtition time limits and hardware failures.
+
```
Comment (ShobanaSekar) [inline docs/source/best-practices/checkpointing.rst:6]: typo on partition time limits

File: docs/source/best-practices/checkpointing.rst [hunk 3]
```diff
@@ -0,0 +1,162 @@
+*****************************************
+Checkpoint/Restart Discovery Jobs
+*****************************************
+
+The complexity of HPC systems may introduce uncontrollable and unpredictable behaviors and may result in job failures due to hardware or software. Applying fault tolerance techniques in your HPC workflows allows your jobs to become more resilient to crashes, prtition time limits and hardware failures.
```
Comment (ShobanaSekar) [inline docs/source/best-practices/checkpointing.rst:5]: can we use another word here for 'uncontrollable'?

File: docs/source/best-practices/checkpointing.rst [hunk 4]
```diff
@@ -0,0 +1,162 @@
...
+ 
+ module load cuda/10.2
+ module load gcc/7.3.0
+ module load openmpi/4.0.5-skylake-gcc7.3
+ module load gromacs/2020.3-gpu-mpi
+ source /shared/centos7/gromacs/2020.3-gcc7.3/bin/GMXRC.bash
+
+ srun --mpi=pmi2 -n $SLURM_NTASKS gmx_mpi mdrun -ntomp 1 -s myrun.tpr -v -dlb yes -cpi state
+
+Where we used the checkpoint flag followed by the file name `-cpi state` to be used to checkpointing. This directs mdrun to use the checkpoint file named `state.cpt` when loading the state. The Slurm option `--array=1-10%1` will create 10 Slurm array tasks, and will run one task job at a time for 12 hours. Note that the saved variable `%A` denotes the main job ID, while variable `%a` denotes the task ID (spanning values 1-10).
```
Comment (ShobanaSekar) [inline docs/source/best-practices/checkpointing.rst:76]: In the above script, we use the checkpoint flag followed by the file name `-cpi state` to be used for checkpointing.

File: docs/source/best-practices/checkpointing.rst [hunk 5]
```diff
@@ -0,0 +1,162 @@
...
+
+To submit these jobs to the scheduler, use the command::
+
+   sbatch submit_mdrun_array.bash
+
+Python TensorFlow Checkpointing example
+~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
+
+This example demonstrates how implement a longer TensorFlow ML training using the **tf.keras** Checkpointing `API <https://www.tensorflow.org/tutorials/keras/save_and_load>`_ and multiple shorter Slurm job arrays on the gpu partition.
```
Comment (ShobanaSekar) [inline docs/source/best-practices/checkpointing.rst:85]: ...demonstrates how **to** implement a longer TensorFlow ML training **job** or **task**...

File: docs/source/best-practices/checkpointing.rst [hunk 6]
```diff
@@ -0,0 +1,162 @@
...
+    verbose=1,
+    save_weights_only=True,
+    period=5)
+
+The full scripts be found `here <https://github.com/NURC-Training/checkpointing-102021/tree/main/Exercise_2>`_ and were modified from `TensorFlow Save and load models <https://www.tensorflow.org/tutorials/keras/save_and_load>`_.
+
+The Slurm option `--array=1-10%1` will create 10 Slurm array tasks, and will run one task job at a time. Note that the saved variable `%A` denotes the main job ID, while variable `%a` denotes the task ID (spanning values 1-10).
+The Shell variable SLURM_ARRAY_TASK_ID denotes the unique task job value. Note that also the output/error files are unique in order to prevent different jobs writing to the same files.
+
+To submit these jobs to the scheduler, use the command::
```
Comment (ShobanaSekar) [inline docs/source/best-practices/checkpointing.rst:123]: To submit **this job** to the scheduler

File: docs/source/best-practices/checkpointing.rst [hunk 7]
```diff
@@ -0,0 +1,162 @@
...
+   
+  sbatch submit_tf_array.bash
+
+Checkpointing using DMTCP
+~~~~~~~~~~~~~~~~~~~~~~~~~~~~
+
+`DMTCP <https://dmtcp.sourceforge.io/>`_ (Distributed MultiThreaded Checkpointing) is a Checkpointing tool that lets you Checkpoint without the need to change your code. It Works with most Linux applications such as Python, Matlab, R, GUI, MPI etc. 
+The program runs in the background of your program, without significant performance loss, and saves the process states into checkpoint files. DMTCP is available on the cluster ::
```
Comment (ShobanaSekar) [inline docs/source/best-practices/checkpointing.rst:131]: you can finish this sentence by saying "and can be accessed as follows:" or something similar

File: docs/source/best-practices/checkpointing.rst [hunk 8]
```diff
@@ -0,0 +1,162 @@
...
+What data to save?
+
+ * Non-temporary application data
+ * Any application data that has been modified since the last checkpoint
+ * Delete checkpoints that are no longer useful - keep only the most recent checkpoint file.
+
+How frequently to checkpoint? 
+
+ * Too often – will slow down your calculation, may be I/O heavy and memory-limited.
+ * Too infrequently – leads to large/long rollbacks times.
```
Comment (ShobanaSekar) [inline docs/source/best-practices/checkpointing.rst:153]: **rollback** instead of rollbacks?

**Review by manasvita — APPROVED** [review]
Overall looks good!
